ssl/ssl_privkey.cc

```diff
@@ -397,6 +397,45 @@ bool ssl_public_key_supports_signature_algorithm(SSL_HANDSHAKE *hs,
   return true;
 }
 
+static UniquePtr<EVP_PKEY> ssl_cert_parse_leaf_pubkey(
+    STACK_OF(CRYPTO_BUFFER) *chain) {
+  CBS leaf;
+  CRYPTO_BUFFER_init_CBS(sk_CRYPTO_BUFFER_value(chain, 0), &leaf);
+  return ssl_cert_parse_pubkey(&leaf);
+}
+
+bool ssl_cert_private_keys_supports_legacy_signature_algorithm(
+    uint16_t *out, SSL_HANDSHAKE *hs) {
+  SSL *const ssl = hs->ssl;
+  assert(ssl_protocol_version(ssl) < TLS1_2_VERSION);
+
+  CERT *cert = hs->config->cert.get();
+  if (cert == nullptr || !ssl->server) {
+    return false;
+  }
+
+  for (size_t i = 0; i < cert->cert_private_keys.size(); i++) {
+    EVP_PKEY *private_key = cert->cert_private_keys[i].privatekey.get();
+    if (private_key != nullptr &&
+        // We may have a private key that supports the signature algorithm,
+        // but we need to verify that the negotiated cipher allows it.
+        hs->new_cipher->algorithm_auth &
+            ssl_cipher_auth_mask_for_key(private_key) &&
+        tls1_get_legacy_signature_algorithm(out, private_key)) {
```

Comment:
Great questions about preferring higher security. There's some things about multiple certificate slots that I should clarify first though. These certificate slots correspond to 1 slot per "key type". This means that we only maintain 1 key at most for RSA, ECDSA, and ED25519 and we won't have multiple security levels of RSA keys or ECDSA keys.

There's not much we can do for TLS1.0/1.1 since there are only two possible signature algorithms, and we're mainly choosing a key that can correspond with the negotiated cipher suite (either RSA or ECDSA). I did take this into account when doing TLS1.2/1.3 however. 
https://github.com/aws/aws-lc/blob/f1d3a58f2e8c33598e8c5e30bbdf1d803db6c7db/ssl/extensions.cc#L4112-L4125

For TLS1.2/1.3, we have a list of signature algorithms ordered in priority (framework already existed prior to my change). We choose the first signature algorithm that any of our private key slots has indicated support for.